flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/LazyBinaryFormat.java

```diff
@@ -68,28 +72,43 @@ public void setJavaObject(T javaObject) {
 
 	@Override
 	public MemorySegment[] getSegments() {
-		ensureMaterialized();
+		if (segments == null) {
+			throw new IllegalStateException("Lazy Binary Format was not materialized");
+		}
 		return segments;
 	}
 
 	@Override
 	public int getOffset() {
-		ensureMaterialized();
+		if (segments == null) {
+			throw new IllegalStateException("Lazy Binary Format was not materialized");
+		}
 		return offset;
 	}
 
 	@Override
 	public int getSizeInBytes() {
-		ensureMaterialized();
+		if (segments == null) {
+			throw new IllegalStateException("Lazy Binary Format was not materialized");
+		}
 		return sizeInBytes;
 	}
 
 	/**
 	 * Ensure we have materialized binary format.
 	 */
-	public void ensureMaterialized() {
+	public final void ensureMaterialized(TypeSerializer<T> serializer) {
 		if (segments == null) {
-			materialize();
+			try {
+				Binary binary = materialize(serializer);
+				this.offset = binary.offset;
+				this.sizeInBytes = binary.sizeInBytes;
+
+				// segments must be set last as this variable tells if the object was materialized
```

Comment:
If we just store a `Binary`, that will be safer.

Reply:
Initially I didn't want to change it this way, because of how the class hierarchy was organized. The problem is that the subclasses access the fields of `BinaryFormat` directly. I tried to clean that up a little bit. Please take a look at the updated PR.